File: js/mergeSort.js
class MergeSort {
    
    constructor(data) {
        this.data = data;  
    }

    merge(st, en, mid) {
        const temp = {}; 
        let lidx = st; 
        let ridx = mid + 1; 
        
       for (let idx = st; idx <= en; idx++) {
            // console.log(this.data, left, right, lidx, ridx)
            if (lidx === mid + 1) temp[idx] = this.data[ridx++]; 
            else if (ridx === en + 1) temp[idx] = this.data[lidx++]; 
            else if (this.data[lidx] <= this.data[ridx]) temp[idx] = this.data[lidx++]; 
            else temp[idx] = this.data[ridx++];
        }

        for (const [idx, value] of Object.entries(temp)) {
            this.data[Number(idx)] = value; 
        }
    }

    sort(st, en) {
        if (st == en) return; 
        let mid = Math.floor((st + en) / 2); 
        this.sort(st, mid); 
        this.sort(mid+1, en); 
        this.merge(st, en, mid); 
    } 

}

const mine = new MergeSort([5, 3, 1, 2])
mine.sort(0, mine.data.length - 1); 
console.log(mine.data); 

File: cpp/merge_sort.cpp
#include <bits/stdc++.h>
using namespace std;

class MergeSort {

public:
    vector<int> data; 
    MergeSort(vector<int> _data): data(_data) {};
    ~MergeSort() = default;

    void sort(int front, int back) {
        if (front == back)
            return;

        int mid = (front + back) / 2;
        sort(front, mid);
        sort(mid + 1, back);

        vector<int> temp{};
        int fidx = front;
        int bidx = mid + 1; 

        while (temp.size() <= (back - front + 1)) {
            int left = data[fidx];
            int right = data[bidx]; 

            if (fidx >= mid + 1) {
                temp.push_back(data[bidx++]); 
                continue;
            }
            if (bidx > back) {
                temp.push_back(data[fidx++]); 
                continue;
            }

            if (left <= right) {
                temp.push_back(data[fidx++]);
            } else {
                temp.push_back(data[bidx++]);
            }
        }

        int cur_idx = front; 
        for (auto& elem : temp) {
            data[cur_idx++] = elem; 
        }

        return;
    }

    void print() {
        for (auto& elem: data) {
            cout << elem << " "; 
        }
        cout << '\n'; 
    }
};

int main() {
    vector<int> data{1, 5, 11, 6, 7, 8, 15, 30, 55, 22, 26};
    MergeSort temp = MergeSort(data);
    temp.sort(0, temp.data.size() - 1);
    temp.print();
    return 0;
}